crates/core/src/client/mod.rs

```diff
@@ -322,26 +333,41 @@ impl<P: Provider + Send + Sync + 'static> KakarotEthApi<P> for KakarotClient<P>
     }
 
     /// Returns the nonce for a given ethereum address
+    /// if it's an EOA, use native nonce and if it's a contract account, use managed nonce
     /// if ethereum -> stark mapping doesn't exist in the starknet provider, we translate
     /// ContractNotFound errors into zeros
     async fn nonce(&self, ethereum_address: Address, block_id: BlockId) -> Result<U256, EthApiError<P::Error>> {
         let starknet_block_id: StarknetBlockId = EthBlockId::new(block_id).try_into()?;
         let starknet_address = self.compute_starknet_address(ethereum_address, &starknet_block_id).await?;
 
-        self.starknet_provider
-            .get_nonce(starknet_block_id, starknet_address)
-            .await
-            .map(|nonce| {
-                let nonce: Felt252Wrapper = nonce.into();
-                nonce.into()
-            })
-            .or_else(|err| match err {
-                ProviderError::StarknetError(StarknetErrorWithMessage {
-                    code: MaybeUnknownErrorCode::Known(StarknetError::ContractNotFound),
-                    ..
-                }) => Ok(U256::from(0)),
-                _ => Err(EthApiError::from(err)),
-            })
+        // Get the implementation of the account
+        let account = KakarotAccount::new(starknet_address, &self.starknet_provider);
+        let implementation = match account.implementation(&starknet_block_id).await {
+            Ok(class_hash) => class_hash,
+            Err(_) => return Ok(U256::from(0)), // Return 0 if the account doesn't exist
```

Comment:
are there other errors that need to bubble up? 
like if account doesn't exist then 0 but what if we fail our call for another reason

Reply:
Right we need to handle `ContractNotFound` and return 0, otherwise we propagate the error, will fix